File: sdk/java/src/main/java/com/anyvali/schemas/NumberSchema.java
package com.anyvali.schemas;

import com.anyvali.IssueCodes;
import com.anyvali.Schema;
import com.anyvali.ValidationContext;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Base class for number/float schemas with numeric constraints.
 * NumberSchema is the number alias (= float64).
 */
public class NumberSchema extends Schema<Double> {
    protected String kind;
    protected Double min;
    protected Double max;
    protected Double exclusiveMin;
    protected Double exclusiveMax;
    protected Double multipleOf;

    public NumberSchema() {
        this.kind = "number";
    }

    protected NumberSchema(String kind) {
        this.kind = kind;
    }

    protected NumberSchema(NumberSchema other) {
        this.coercion = other.coercion;
        this.defaultValue = other.defaultValue;
        this.hasDefault = other.hasDefault;
        this.kind = other.kind;
        this.min = other.min;
        this.max = other.max;
        this.exclusiveMin = other.exclusiveMin;
        this.exclusiveMax = other.exclusiveMax;
        this.multipleOf = other.multipleOf;
    }

    public NumberSchema min(double v) {
        var copy = copyAs();
        copy.min = v;
        return copy;
    }

    public NumberSchema max(double v) {
        var copy = copyAs();
        copy.max = v;
        return copy;
    }

    public NumberSchema exclusiveMin(double v) {
        var copy = copyAs();
        copy.exclusiveMin = v;
        return copy;
    }

    public NumberSchema exclusiveMax(double v) {
        var copy = copyAs();
        copy.exclusiveMax = v;
        return copy;
    }

    public NumberSchema multipleOf(double v) {
        var copy = copyAs();
        copy.multipleOf = v;
        return copy;
    }

    protected NumberSchema copyAs() {
        return new NumberSchema(this);
    }

    @Override
    protected Object validate(Object input, ValidationContext ctx) {
        if (!checkIsNumber(input, ctx)) {
            return null;
        }

        double value = toDouble(input);
        checkRange(value, ctx);
        checkConstraints(value, ctx);
        return value;
    }

    protected boolean checkIsNumber(Object input, ValidationContext ctx) {
        if (input instanceof Boolean) {
            ctx.addIssue(IssueCodes.INVALID_TYPE,
                    "Expected " + kind + ", received boolean", kind, "boolean");
            return false;
        }
        if (!(input instanceof Number)) {
            String received = anyvaliTypeName(input);
            ctx.addIssue(IssueCodes.INVALID_TYPE,
                    "Expected " + kind + ", received " + received, kind, received);
            return false;
        }
        double d = ((Number) input).doubleValue();
        if (Double.isNaN(d) || Double.isInfinite(d)) {
            ctx.addIssue(IssueCodes.INVALID_NUMBER,
                    "Number must be finite", "finite number", input);
            return false;
        }
        return true;
    }

    protected void checkRange(double value, ValidationContext ctx) {
        // Overridden in Float32Schema
    }

    protected void checkConstraints(double value, ValidationContext ctx) {
        if (min != null && value < min) {
            ctx.addIssue(IssueCodes.TOO_SMALL,
                    "Number must be >= " + min, min, value);
        }
        if (max != null && value > max) {
            ctx.addIssue(IssueCodes.TOO_LARGE,
                    "Number must be <= " + max, max, value);
        }
        if (exclusiveMin != null && value <= exclusiveMin) {
            ctx.addIssue(IssueCodes.TOO_SMALL,
                    "Number must be > " + exclusiveMin, exclusiveMin, value);
        }
        if (exclusiveMax != null && value >= exclusiveMax) {
            ctx.addIssue(IssueCodes.TOO_LARGE,
                    "Number must be < " + exclusiveMax, exclusiveMax, value);
        }
        if (multipleOf != null && multipleOf != 0) {
            if (value % multipleOf != 0) {
                ctx.addIssue(IssueCodes.INVALID_NUMBER,
                        "Number must be a multiple of " + multipleOf,
                        multipleOf, value);
            }
        }
    }

    @Override
    protected Map<String, Object> toNode() {
        var node = new LinkedHashMap<String, Object>();
        node.put("kind", kind);
        if (min != null) node.put("min", min);
        if (max != null) node.put("max", max);
        if (exclusiveMin != null) node.put("exclusiveMin", exclusiveMin);
        if (exclusiveMax != null) node.put("exclusiveMax", exclusiveMax);
        if (multipleOf != null) node.put("multipleOf", multipleOf);
        return addCommonNodeFields(node);
    }

    @Override
    protected Schema<Double> copy() {
        return new NumberSchema(this);
    }

    protected static double toDouble(Object input) {
        return ((Number) input).doubleValue();
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/schemas/NumberSchema.kt
package com.anyvali.schemas

import com.anyvali.*
import com.anyvali.parse.Coercion
import kotlinx.serialization.json.*

data class NumberSchema(
    val schemaKind: String = "number",
    val min: Double? = null,
    val max: Double? = null,
    val exclusiveMin: Double? = null,
    val exclusiveMax: Double? = null,
    val multipleOf: Double? = null,
    val defaultValue: Any? = UNSET,
    val coerce: String? = null
) : Schema<Double>() {
    override val kind: String = schemaKind

    fun min(n: Number) = copy(min = n.toDouble())
    fun max(n: Number) = copy(max = n.toDouble())
    fun exclusiveMin(n: Number) = copy(exclusiveMin = n.toDouble())
    fun exclusiveMax(n: Number) = copy(exclusiveMax = n.toDouble())
    fun multipleOf(n: Number) = copy(multipleOf = n.toDouble())
    fun default(v: Number) = copy(defaultValue = v)
    fun coerce(c: String) = copy(coerce = c)
    /**
     * No-arg ergonomic: enable string coercion with the target inferred from the
     * schema kind. Equivalent to the portable generic "string" source (spec 5.1)
     * and to JS `.coerce()` / `.coerce({ from: "string" })`.
     */
    fun coerce() = copy(coerce = "string")

    fun describe(description: String, opts: DescribeOptions? = null): NumberSchema {
        applyDescribe(description, opts)
        return this
    }

    fun metadata(meta: Map<String, Any?>, replace: Boolean = false): NumberSchema {
        applyMetadata(meta, replace)
        return this
    }

    @Suppress("UNCHECKED_CAST")
    override fun safeParseWithContext(input: Any?, ctx: ValidationContext): ParseResult<Double> {
        var value = input

        // Apply coercion. Both the typed token ("string->number") and the generic
        // portable "string" source (set via the no-arg coerce()) enable
        // string->number coercion, with the target inferred from this schema's kind.
        if ((coerce == "string->number" || coerce == "string") && value is String) {
            val parsed = Coercion.coerceStringToNumber(value)
            if (parsed == null) {
                return ParseResult.Failure(
                    listOf(
                        ValidationIssue(
                            code = IssueCodes.COERCION_FAILED,
                            path = ctx.path,
                            expected = kind,
                            received = value
                        )
                    )
                )
            }
            value = parsed
        }

        val issues = validateValue(value, ctx)
        return if (issues.isEmpty()) {
            ParseResult.Success((value as Number).toDouble())
        } else {
            ParseResult.Failure(issues)
        }
    }

    override fun validateValue(value: Any?, ctx: ValidationContext): List<ValidationIssue> {
        if (value == null || value is Boolean || value is String || value is List<*> || value is Map<*, *>) {
            return listOf(
                ValidationIssue(
                    code = IssueCodes.INVALID_TYPE,
                    path = ctx.path,
                    expected = kind,
                    received = getJsonTypeName(value)
                )
            )
        }
        if (value !is Number) {
            return listOf(
                ValidationIssue(
                    code = IssueCodes.INVALID_TYPE,
                    path = ctx.path,
                    expected = kind,
                    received = getJsonTypeName(value)
                )
            )
        }

        val dbl = toDouble(value)
        val issues = mutableListOf<ValidationIssue>()

        if (dbl.isNaN() || dbl.isInfinite()) {
            issues.add(
                ValidationIssue(
                    code = IssueCodes.INVALID_NUMBER,
                    path = ctx.path,
                    expected = kind,
                    received = dbl.toString()
                )
            )
            return issues
        }

        // Float32 range check
        if (schemaKind == "float32") {
            if (dbl != 0.0 && (dbl > Float.MAX_VALUE.toDouble() || dbl < -Float.MAX_VALUE.toDouble())) {
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.TOO_LARGE,
                        path = ctx.path,
                        expected = "float32",
                        received = dbl.toString()
                    )
                )
            }
        }

        min?.let {
            if (dbl < it) {
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.TOO_SMALL,
                        path = ctx.path,
                        expected = formatNumber(it),
                        received = formatNumber(dbl)
                    )
                )
            }
        }
        max?.let {
            if (dbl > it) {
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.TOO_LARGE,
                        path = ctx.path,
                        expected = formatNumber(it),
                        received = formatNumber(dbl)
                    )
                )
            }
        }
        exclusiveMin?.let {
            if (dbl <= it) {
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.TOO_SMALL,
                        path = ctx.path,
                        expected = formatNumber(it),
                        received = formatNumber(dbl)
                    )
                )
            }
        }
        exclusiveMax?.let {
            if (dbl >= it) {
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.TOO_LARGE,
                        path = ctx.path,
                        expected = formatNumber(it),
                        received = formatNumber(dbl)
                    )
                )
            }
        }
        multipleOf?.let {
            val remainder = dbl % it
            if (Math.abs(remainder) > 1e-10 && Math.abs(remainder - it) > 1e-10) {
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.INVALID_NUMBER,
                        path = ctx.path,
                        expected = formatNumber(it),
                        received = formatNumber(dbl)
                    )
                )
            }
        }

        return issues
    }

    override fun exportNode(): JsonObject = buildJsonObject {
        put("kind", JsonPrimitive(schemaKind))
        min?.let { put("min", JsonPrimitive(it)) }
        max?.let { put("max", JsonPrimitive(it)) }
        exclusiveMin?.let { put("exclusiveMin", JsonPrimitive(it)) }
        exclusiveMax?.let { put("exclusiveMax", JsonPrimitive(it)) }
        multipleOf?.let { put("multipleOf", JsonPrimitive(it)) }
        if (defaultValue !== UNSET) {
            val dv = defaultValue
            when (dv) {
                is Double -> put("default", JsonPrimitive(dv))
                is Int -> put("default", JsonPrimitive(dv))
                is Long -> put("default", JsonPrimitive(dv))
                is Float -> put("default", JsonPrimitive(dv))
                else -> {}
            }
        }
        coerce?.let { put("coerce", JsonPrimitive(it)) }
        addMetadataToNode(this)
    }

    companion object {
        val UNSET = Any()

        fun formatNumber(d: Double): String {
            return if (d == d.toLong().toDouble() && !d.isNaN() && !d.isInfinite()) {
                d.toLong().toString()
            } else {
                d.toString()
            }
        }
    }
}
